src/document_change_manager.ts

```diff
@@ -269,6 +269,9 @@ export class DocumentChangeManager implements Disposable {
                         },
                         { undoStopAfter: false, undoStopBefore: false },
                     );
+                    // Although successful, the document version may not actually increase.
+                    // Example: Using "S" to delete empty line does not actually change the text.
+                    this.documentSkipVersionOnChange.set(doc, doc.version);
```

Comment:
I haven't looked at this cod before, so forgive the silly question, but is there any reason to do the `set` on like 260, still?

Reply:
It's indeed a question; I am Not sure what the event-triggering cycle of VSCode is like.

Reply:
Hm, upon reflection, I guess the event handler for a document being changed could be fired before we add this to the map. We'd need some kind of mutual exclusion here, I think.

Reply:
> We'd need some kind of mutual exclusion here

It seems a bit tricky. How do we handle the issue of outdated data? We need to maintain the sequence of events triggered by two completely different operations. Storing the wrong version number here could result in losing changes made next time in VS Code. Correcting the version number after applying changes should work.

Reply:
It does. Though I do think there's some potential for races regardless. I think it would make some sense to have a lock shared between both syncs, so that only one sync direction is handled at a time. That said, I don't think its necessary for this PR per-se.

Reply:
Using locks seems to disturb the event sequence instead. 

If both directions simultaneously undergo different changes, then the content is directly wrong.